### PR TITLE
[tests][dask] test all boosting types (fixes #3896)

### LINKED ISSUE #3896: [dask] Support all LightGBM boosting types
## Summary

Right now the Dask interface in https://github.com/microsoft/LightGBM/blob/706f2af7badc26f6ec68729469ec6ec79a66d802/python-package/lightgbm/dask.py  is tested only with `gbdt` (default) boosting type.
https://lightgbm.readthedocs.io/en/latest/Parameters.html#boosting

I believe that parametrization of core tests (`test_classifier`, `test_regressor`, `test_ranker`) with different boosting types (`rf`, `dart`, `goss`) will improve tests and make us confident in Dask module quality.
Please pay attention to that for some boosting types some parameters cannot be left with their default values. For example, `rf` performs the following checks
https://github.com/microsoft/LightGBM/blob/fcfd4132e6d40a22d52023396329c41fd3de4a42/src/boosting/rf.hpp#L35-L36

How tests can be parametrized:
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L31
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L166

### PR DESCRIPTION
**Pull request by jmoralez**
This includes all boosting types (gbdt, rf, goss, dart) in `test_classifier`, `test_regressor` and `test_ranker`

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
This looks great, thanks so much for the contribution amd for documenting the issues you faced with random forest mode in #4118. @StrikerRUS are you ok with this increase in the number of tests? I think these are providing valuable new coverage and are worth it, but want to get your opinion.

File: tests/python_package_test/test_dask.py [hunk 1]
```diff
@@ -263,13 +273,17 @@ def test_classifier(output, task, client):
...
-    assert_eq(y, p2)
-    assert_eq(p1_proba, p2_proba, atol=0.01)
-    assert_eq(p1_local, p2)
-    assert_eq(y, p1_local)
+    if boosting_type == 'rf' and output == 'dataframe-with-categorical':
+        # https://github.com/microsoft/LightGBM/issues/4118
+        assert_eq(s1, s2, atol=0.01)
+        assert_eq(p1_proba, p2_proba, atol=0.8)
+    else:
+        assert_eq(p1, p2)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:281]: Why are we dropping `assert_eq(s1, s2)` from the `else` branch?

**Review by StrikerRUS — COMMENTED** [review]
> @StrikerRUS are you ok with this increase in the number of tests? I think these are providing valuable new coverage and are worth it, but want to get your opinion.

Thank you. I support this and think this is valuable.
Just one question below.

File: tests/python_package_test/test_dask.py (hunk 1 above)
Comment (jmoralez) [reply tests/python_package_test/test_dask.py:282]: That was totally my bad, I lost it when splitting the statements. I've included it in 
5d7001d

Comment (jameslamb) [reply tests/python_package_test/test_dask.py:282]: thanks for catching it @StrikerRUS ! I missed it too

**Review by StrikerRUS — APPROVED** [review]
Great, many thanks!

**Pull request merged**